doc/operations/code_upgrade.rst

```diff
@@ -0,0 +1,189 @@
+Code Upgrade
+============
+
+.. note:: Refer to :doc:`/operations/code_upgrade_1x` for specific guidelines on how to upgrade a 1.x CCF service to 2.0.
+
+This page describes how operators/members can upgrade a live CCF service to a new version with minimal downtime.
+
+Reasons for running the code upgrade procedure include:
+
+- Upgrading nodes to a new version of a C++ application or JavaScript runtime (i.e. ``libjs_generic.enclave.so.signed``).
+- Upgrading nodes to a new version of CCF.
+
+.. tip::
+
+    - Note that there is no need to run the code upgrade procedure detailed on this page if `only` the JavaScript/TypeScript application needs updating (see :ref:`JavaScript/TypeScript bundle deployment procedure <build_apps/js_app_bundle:Deployment>`).
+    - If more than a majority of nodes have failed, the disaster recovery procedure should be run by operators instead (see :doc:`/operations/recovery`).
+
+.. note:: CCF guarantees specific live compatibility across different LTS versions. See :ref:`overview/release_policy:Operations compatibility` for more details.
+
+Procedure
+---------
+
+1. Let's assume that the to-be-upgraded service is made of 3 nodes (tolerates up to one fault, i.e. ``f = 1``), with ``Node 1`` as the primary node (the code upgrade procedure can be run from any number of nodes):
+
+.. mermaid::
+
+    graph TB;
+        classDef Primary stroke-width:4px
+
+        subgraph Service
+            Node0((Node 0))
+            Node1((Node 1))
+            class Node1 Primary
+            Node2((Node 2))
+        end
+
+1. First, operators/members should register the new code version corresponding to the new enclave measurement using the ``add_node_code`` proposal action (see :ref:`governance/common_member_operations:Updating Code Version`).
+
+2. The set of new nodes running the enclave registered in the previous step should be added to the service (see :ref:`operations/start_network:Adding a New Node to the Network`) and trusted by members (see :ref:`governance/common_member_operations:Trusting a New Node`). Typically, the same number of nodes than were originally present should be added to the service. In this example, the service is now made of 6 nodes (``f = 2``).
+
+.. mermaid::
+
+    graph TB;
+        classDef NewNode fill:turquoise
+        classDef Primary stroke-width:4px
+
+        subgraph Service
+            subgraph Old Nodes
+                Node0((Node 0))
+                Node1((Node 1))
+                class Node1 Primary
+                Node2((Node 2))
+            end
+
+            subgraph New Nodes
+                Node3((Node 3))
+                Node4((Node 4))
+                Node5((Node 5))
+                class Node3 NewNode
+                class Node4 NewNode
+                class Node5 NewNode
+            end
+        end
+
+
+3. The old code version (i.e. the code version run by nodes 0, 1 and 2) can be removed using the ``remove_node_code`` proposal action.
```

Comment:
Doing the code version swap atomically should be fine, except that it would need to coordinate with the automated node re-provisioning service: if old nodes need auto-replacement during the transition, that can only succeed if the old code id is removed.

Depending how the auto-provisioning works, keeping the old code id in the table until quite late may be necessary.

Reply:
Good point. I'd moved this further down than the order in which we do it in the end-to-end test but it's safer to move it to the latest stage.